src/internal/pjs/api_server.go

```diff
@@ -183,38 +199,46 @@ func (a *apiServer) ProcessQueue(srv pjs.API_ProcessQueueServer) (retErr error)
 			return err
 		}
 		var (
-			inputHandles []*fileset.Handle
-			inputHashes  [][]byte
-			programHash  []byte
+			programHash []byte
+			inputs      []fileset.Pin
+			inputHashes [][]byte
 		)
-		if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
-			job, err := pjsdb.GetJob(ctx, sqlTx, jobID)
+		if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, tx *pachsql.Tx) error {
+			job, err := pjsdb.GetJob(ctx, tx, jobID)
 			if err != nil {
 				return errors.Wrap(err, "get job")
 			}
 			programHash = job.ProgramHash
 			for _, input := range job.Inputs {
-				inputHandles = append(inputHandles, fileset.NewHandle(input))
+				inputs = append(inputs, input)
 			}
 			return nil
 		}, dbutil.WithReadOnly()); err != nil {
 			return errors.Wrap(err, "with tx")
 		}
-		var inputs []string
-		for i, handle := range inputHandles {
-			inputs = append(inputs, handle.HexString())
-			hash, err := HashFileset(ctx, a.env.GetStorageClient(ctx), handle.HexString())
-			if err != nil {
-				return errors.Wrapf(err, "hashing input fileset: %q, (%d/%d)", handle.HexString(), i, len(inputHandles))
+		var inputHandles []string
+		for _, input := range inputs {
+			if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, tx *pachsql.Tx) error {
+				// TODO: TTL?
+				handle, err := a.env.Storage.Filesets.GetPinHandleTx(ctx, tx, input, 15*time.Minute)
+				if err != nil {
+					return err
+				}
+				inputHandles = append(inputHandles, handle.HexString())
+				id := handle.ID()
+				inputHashes = append(inputHashes, id[:])
+				return nil
+			}); err != nil {
+				return err
 			}
-			inputHashes = append(inputHashes, hash)
 		}
 		if err := srv.Send(&pjs.ProcessQueueResponse{
 			Context: hex.EncodeToString(jobCtx.Token),
-			Input:   inputs,
+			Input:   inputHandles,
 		}); err != nil {
 			return errors.Wrap(err, "send")
 		}
+		// TODO: Shouldn't there be code here to disconnect if the job is deleted / canceled?
```

Comment:
I have to re-examine this code again more carefully, but you may get another job on the queue after the deleted/cancelled one.

Reply:
I don't think it works like that since it just waits on a message from the worker here. We would need to communicate to the worker that it needs to stop working on the current job. What was suggested in the original design was to disconnect from the worker (i.e. error out of the RPC), then workers would generally be built to retry when this happens. This would ensure they stop working on a job as soon as it is deleted / canceled.

Reply:
Gotcha, Let's leave this TODO. We'll can create a story for this and the add the code to disconnect plus a test to make sure that new code works.